firebase-firestore/src/test/java/com/google/firebase/firestore/local/IndexBackfillerTest.java

```diff
@@ -276,6 +287,129 @@ public void testBackfillUsesLatestReadTimeForEmptyCollections() {
+    FieldIndex fieldIndex = indexManager.getFieldIndexes("coll1").iterator().next();
+    assertEquals(2, fieldIndex.getIndexState().getOffset().getLargestBatchId());
+
+    documentsProcessed = backfiller.backfill();
+    assertEquals(2, documentsProcessed);
+    verifyQueryResults("coll1", "coll1/docA", "coll1/docB", "coll1/docC", "coll1/docD");
+    fieldIndex = indexManager.getFieldIndexes("coll1").iterator().next();
+    assertEquals(4, fieldIndex.getIndexState().getOffset().getLargestBatchId());
+  }
+
+  // TODO(indexing): Support adding local mutations to collection group table so this test passes.
+  @Test
+  @Ignore
+  public void testBackfillMutationsWithNoPriorRemoteDoc() {
+    addFieldIndex("coll1", "foo");
+    addSetMutationsToOverlay(2, "coll1/docA");
+    addSetMutationsToOverlay(4, "coll1/docB");
+
+    int documentsProcessed = backfiller.backfill();
+    assertEquals(2, documentsProcessed);
+  }
+
+  @Test
+  public void testBackfillMutationFinishesMutationBatchEvenIfItExceedsLimit() {
+    backfiller.setMaxDocumentsToProcess(2);
+    addFieldIndex("coll1", "foo");
+    addDoc("coll1/docA", "foo", version(10));
+    addSetMutationsToOverlay(2, "coll1/docB", "coll1/docC", "coll1/docD");
+    addSetMutationsToOverlay(3, "coll1/docE");
+
+    int documentsProcessed = backfiller.backfill();
+    assertEquals(4, documentsProcessed);
+    verifyQueryResults("coll1", "coll1/docA", "coll1/docB", "coll1/docC", "coll1/docD");
+  }
+
+  @Test
+  public void testBackfillMutationsFromHighWaterMark() {
+    backfiller.setMaxDocumentsToProcess(2);
+    addFieldIndex("coll1", "foo");
+    addDoc("coll1/docA", "foo", version(10));
+    addSetMutationsToOverlay(3, "coll1/docB");
+
+    int documentsProcessed = backfiller.backfill();
+    assertEquals(2, documentsProcessed);
+    verifyQueryResults("coll1", "coll1/docA", "coll1/docB");
+
+    addSetMutationsToOverlay(1, "coll1/docC");
+    addSetMutationsToOverlay(2, "coll1/docD");
+    documentsProcessed = backfiller.backfill();
+    assertEquals(0, documentsProcessed);
+  }
+
+  @Test
+  public void testBackfillUpdatesExistingDocToNewValue() {
+    backfiller.setMaxDocumentsToProcess(1);
+    addFieldIndex("coll1", "foo");
+    FieldIndex fieldIndex = indexManager.getFieldIndexes("coll1").iterator().next();
+
+    Document docA = addDoc("coll1/docA", "foo", version(10));
+    int documentsProcessed = backfiller.backfill();
+    assertEquals(1, documentsProcessed);
+    SortedSet<IndexEntry> before = indexManager.getExistingIndexEntries(docA.getKey(), fieldIndex);
+
+    // Update doc to new remote version with new value.
+    removeDoc(docA);
+    addDoc("coll1/docA", "foo", version(40), 5);
+    documentsProcessed = backfiller.backfill();
+    assertEquals(1, documentsProcessed);
+    SortedSet<IndexEntry> after = indexManager.getExistingIndexEntries(docA.getKey(), fieldIndex);
+    assertNotEquals(before.first().getDirectionalValue(), after.first().getDirectionalValue());
```

Comment:
Instead of relying on getExistingIndexEntries, can you just the public API and invoke `getDocumentsMatchingTarget` with two different targets (foo > 10 and foo > 20 for example). Before the update, you can verify that one of them matches and then after the update two match.

Same comment applies below.

Reply:
Done.